File: sshlib/src/test/java/com/trilead/ssh2/AsyncSSHCompatibilityTest.java
package com.trilead.ssh2;

import org.apache.commons.io.IOUtils;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.junit.ClassRule;
import org.junit.Rule;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.testcontainers.containers.GenericContainer;
import org.testcontainers.containers.output.Slf4jLogConsumer;
import org.testcontainers.containers.wait.strategy.LogMessageWaitStrategy;
import org.testcontainers.images.builder.ImageFromDockerfile;

import java.io.IOException;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

/**
 * Integration tests against OpenSSH.
 *
 * @author Kenny Root
 */
public class AsyncSSHCompatibilityTest {
	private static final Logger logger = LoggerFactory.getLogger(AsyncSSHCompatibilityTest.class.getSimpleName());

	@Rule
	public SshLogger sshLogger = new SshLogger(logger);

	private static final String USERNAME = "user123";
	private static final String PASSWORD = "secretpw";

	@ClassRule
	public static GenericContainer server;

	static {
		ImageFromDockerfile baseImage = new ImageFromDockerfile()
				.withFileFromClasspath("requirements.txt", "asyncssh-server/requirements.txt")
				.withFileFromClasspath("server.py", "asyncssh-server/server.py")
				.withFileFromClasspath("Dockerfile", "asyncssh-server/Dockerfile");
		for (String key : PubkeyConstants.KEY_NAMES) {
			baseImage.withFileFromClasspath(key, "com/trilead/ssh2/crypto/" + key);
		}

		server = new GenericContainer(baseImage)
				.withLogConsumer(new Slf4jLogConsumer(logger).withPrefix("DOCKER"))
				.waitingFor(new LogMessageWaitStrategy()
						.withRegEx(".*LISTENER READY.*\\s"));
	}

	@NotNull
	@Contract("_ -> new")
	private Connection withServer(@NotNull GenericContainer container) {
		return new Connection(container.getContainerIpAddress(), container.getMappedPort(8022));
	}

	private ConnectionInfo assertCanPasswordAuthenticate(GenericContainer server) throws IOException {
		try (Connection c = withServer(server)) {
			c.connect();
			assertThat(c.authenticateWithPassword(USERNAME, PASSWORD), is(true));
			try (Session s = c.openSession()) {
				s.ping();
			}
			return c.getConnectionInfo();
		}
	}

	@Test
	public void canConnectWithPassword() throws Exception {
		assertCanPasswordAuthenticate(server);
	}

	private ConnectionInfo assertCanPubkeyAuthenticate(GenericContainer server, char[] key) throws IOException {
		try (Connection c = withServer(server)) {
			c.connect();
			assertThat(c.authenticateWithPublicKey(USERNAME, key, ""), is(true));
			try (Session s = c.openSession()) {
				s.ping();
			}
			return c.getConnectionInfo();
		}
	}

	private void canConnectWithPubkey(String keyFilename) throws Exception {
		char[] keyChars = IOUtils.toCharArray(getClass().getResourceAsStream("crypto/" + keyFilename), "UTF-8");
		assertCanPubkeyAuthenticate(server, keyChars);
	}

	@Test
	public void canConnectWithEd25519() throws Exception {
		canConnectWithPubkey("ed25519-openssh2-private-key.txt");
	}

	@Test
	public void canConnectWithEcdsa256() throws Exception {
		canConnectWithPubkey("ecdsa-nistp256-openssh2-private-key.txt");
	}

	@Test
	public void canConnectWithEcdsa384() throws Exception {
		canConnectWithPubkey("ecdsa-nistp384-openssh2-private-key.txt");
	}

	@Test
	public void canConnectWithEcdsa521() throws Exception {
		canConnectWithPubkey("ecdsa-nistp521-openssh2-private-key.txt");
	}

	@Test
	public void canConnectWithRsa() throws Exception {
		canConnectWithPubkey("rsa-openssh2-private-key.txt");
	}
}

File: sshlib/src/test/kotlin/org/connectbot/sshlib/client/AsyncSshCompatibilityTest.kt
/*
 * Copyright 2025 Kenny Root
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.connectbot.sshlib.client

import org.connectbot.sshlib.HostKeyVerifier
import org.connectbot.sshlib.PublicKey
import org.connectbot.sshlib.SshClientConfig
import org.connectbot.sshlib.blocking.BlockingSshClient
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource
import org.slf4j.LoggerFactory
import org.testcontainers.containers.GenericContainer
import org.testcontainers.containers.output.Slf4jLogConsumer
import org.testcontainers.containers.wait.strategy.LogMessageWaitStrategy
import org.testcontainers.images.builder.ImageFromDockerfile
import org.testcontainers.junit.jupiter.Container
import org.testcontainers.junit.jupiter.Testcontainers

@Testcontainers
class AsyncSshCompatibilityTest {

    companion object {
        private val logger = LoggerFactory.getLogger(AsyncSshCompatibilityTest::class.java)
        private val logConsumer = Slf4jLogConsumer(logger).withPrefix("DOCKER")

        private const val USERNAME = "testuser"
        private const val PASSWORD = "testpass"

        private val PUBLIC_KEY_NAMES = arrayOf(
            "ed25519_unencrypted.pub",
            "ecdsa256_unencrypted.pub",
            "ecdsa384_unencrypted.pub",
            "ecdsa521_unencrypted.pub",
            "rsa_unencrypted.pub"
        )

        private val baseImage = ImageFromDockerfile("asyncssh-server-test", false)
            .withFileFromClasspath(".", "asyncssh-server").also { image ->
                for (key in PUBLIC_KEY_NAMES) {
                    image.withFileFromClasspath(key, "keys/$key")
                }
            }

        @Container
        @JvmStatic
        val server: GenericContainer<*> = GenericContainer(baseImage)
            .withExposedPorts(8022)
            .withLogConsumer(logConsumer)
            .waitingFor(LogMessageWaitStrategy().withRegEx(".*LISTENER READY.*\\s"))

        @JvmStatic
        fun encryptionAlgorithms() = listOf(
            "chacha20-poly1305@openssh.com",
            "aes128-gcm@openssh.com",
            "aes256-gcm@openssh.com",
            "aes128-ctr",
            "aes256-ctr",
            "aes128-cbc",
            "aes256-cbc",
            "3des-cbc"
        )

        @JvmStatic
        fun kexAlgorithms() = listOf(
            "curve25519-sha256",
            "ecdh-sha2-nistp256",
            "ecdh-sha2-nistp384",
            "ecdh-sha2-nistp521",
            "diffie-hellman-group18-sha512",
            "diffie-hellman-group16-sha512",
            "diffie-hellman-group-exchange-sha256",
            "diffie-hellman-group14-sha256",
            "diffie-hellman-group14-sha1",
            "diffie-hellman-group-exchange-sha1",
            "diffie-hellman-group1-sha1"
        )

        @JvmStatic
        fun macAlgorithms() = listOf(
            "hmac-sha2-256-etm@openssh.com",
            "hmac-sha2-512-etm@openssh.com",
            "hmac-sha2-256",
            "hmac-sha2-512",
            "hmac-sha1-etm@openssh.com",
            "hmac-sha1"
        )

        @JvmStatic
        fun pubkeyFiles() = listOf(
            "ed25519_unencrypted",
            "ecdsa256_unencrypted",
            "ecdsa384_unencrypted",
            "ecdsa521_unencrypted",
            "rsa_unencrypted"
        )
    }

    private val acceptAllVerifier = object : HostKeyVerifier {
        override suspend fun verify(key: PublicKey): Boolean = true
    }

    private fun connectToServer(configure: (SshClientConfig.Builder.() -> Unit)? = null): BlockingSshClient {
        val host = server.host
        val port = server.getMappedPort(8022)

        val config = SshClientConfig {
            this.host = host
            this.port = port
            this.hostKeyVerifier = acceptAllVerifier
            configure?.invoke(this)
        }
        val client = BlockingSshClient(config)
        assertTrue(client.connect(), "Should connect to AsyncSSH server")
        return client
    }

    private fun authenticateWithPassword(configure: (SshClientConfig.Builder.() -> Unit)? = null) {
        val client = connectToServer(configure)
        try {
            assertTrue(
                client.authenticatePassword(USERNAME, PASSWORD),
                "Should authenticate with password"
            )
        } finally {
            client.disconnect()
        }
    }

    @Test
    fun `can connect with password`() {
        authenticateWithPassword()
    }

    @ParameterizedTest(name = "pubkey: {0}")
    @MethodSource("pubkeyFiles")
    fun `can connect with public key`(keyFilename: String) {
        val keyData = javaClass.getResourceAsStream("/keys/$keyFilename")!!
            .bufferedReader().readText()
        val client = connectToServer()
        try {
            assertTrue(
                client.authenticatePublicKey(USERNAME, keyData),
                "Should authenticate with $keyFilename"
            )
        } finally {
            client.disconnect()
        }
    }

    @ParameterizedTest(name = "kex: {0}")
    @MethodSource("kexAlgorithms")
    fun `can connect with kex algorithm`(algorithm: String) {
        authenticateWithPassword {
            kexAlgorithms = "$algorithm,kex-strict-c-v00@openssh.com"
        }
    }

    @ParameterizedTest(name = "encryption: {0}")
    @MethodSource("encryptionAlgorithms")
    fun `can connect with encryption algorithm`(algorithm: String) {
        authenticateWithPassword {
            encryptionAlgorithms = algorithm
        }
    }

    @ParameterizedTest(name = "mac: {0}")
    @MethodSource("macAlgorithms")
    fun `can connect with mac algorithm`(algorithm: String) {
        authenticateWithPassword {
            encryptionAlgorithms = "aes128-ctr"
            macAlgorithms = algorithm
        }
    }
}
